README.md
# ShadowProtect Backup Email Processing and Dashboard

ShadowProtect is a great backup solution however determining what that status of all your backups can be a challenge.
What this project does is process all your ShadowProtect emails and generates a dashboard. 
In addition to telling you if a backup completed or failed it can detect whether a backup is an unkown state. 

Requirements
- Debian Linux Server 
- Postfix 
- Fetchmail
- Apache 
- Python

## Theory Of Operation
1. ShadowProtect sends out emails to a dedicated email account.
2. Fetchmail downloads the emails and forwards it locally to postfix.
3. Postfix process the email using a postfix filter/hook which copies the email into a text file stored in a directory.
4. Python script process emails and generates a web page.

## Admin Commands
- To add or remove clients from the system you need to send an email to the email being used by fetchmail with the following structure:
1. To ADD: 
```
Subject: COMMAND

Body:
COMMAND | ADD | Company Name | Backup Client Name
```

2. To REMOVE: 
```
Subject: COMMAND

Body:
COMMAND | REMOVE | Company Name | Backup Client Name
```


ReleaseNotes.md
# Release Notes
## 1.0 Version 12/10/2018 : 
- First Release built on the LAMP Stack

## 2.0 Version 01/09/2019 : 
### What's New In 2.0 Version
- Project has been converted from PHP 7 to Python 3.7 
- MySQL has been replaced by a json file stored on disk
- Removed ini file

## 3.0 Version 01/10/2019 :
### New Features
- Added ability to add or remove backup clients via email





email_commands.py
import re
import json
import glob
import os
import pendulum


class EmailFiles:
    @staticmethod
    def get_list_of_files(pattern):
        return glob.glob(pattern)


class EmailCommands:
    commands = []

    def __init__(self, directory):
        list_of_email_files = EmailFiles.get_list_of_files(directory)
        list_of_command_emails = self.check_for_command_email(list_of_email_files)
        self.commands = self.get_commands(list_of_command_emails)

    @staticmethod
    def check_for_command_email(list_of_email_files):
        command_emails = []
        for email in list_of_email_files:
            with open(email) as f:
                if 'Subject: COMMAND' in f.read():
                    command_emails.append(email)
        return command_emails

    @staticmethod
    def get_commands(list_of_command_emails):
        commands = []
        for email in list_of_command_emails:
            with open(email) as f:
                file_data = f.readlines()
                for line in file_data:
                    if re.search("^COMMAND", line):
                        commands.append(line)
            os.remove(email)
        return commands

    def process_commands(self, dictionary, default_threshold, default_date_format, dictionary_file):
        for command in self.commands:
            c = command.split('|')
            c_command = c[1].strip()
            c_company = c[2].strip()
            c_client = c[3].strip()
            dictionary = self.update_dictionary(c_command, c_company, c_client, dictionary, default_threshold,
                                                default_date_format)
        self.save_json(dictionary, dictionary_file)

    @staticmethod
    def update_dictionary(command, company, client, dictionary, default_threshold, default_date_format):
        now = pendulum.now('US/Eastern')
        threshold_time = now.subtract(hours=default_threshold + 1)
        default_time = threshold_time.strftime(default_date_format)
        key = company + client + client
        if command.upper() == 'REMOVE':
            if key in dictionary.keys():
                del dictionary[key]
        elif command.upper() == 'ADD':
            if key not in dictionary.keys():
                dictionary[key] = {'server': client, 'client': client, 'company': company,
                                   'backup_code': '1120', 'email_time': default_time,
                                   'threshold': default_threshold}
        return dictionary

    @staticmethod
    def save_json(data, file_name):
        with open(file_name, 'w') as fp:
            json.dump(data, fp)


processEmailsClass.php
<?php

class processEmails
{

  private $emailDirectory = '/tmp/email_messages';
  private $dbCredentialsIni;
  private $emailFileNames = [];
  private $emailContents = [];
  private $emailParsedData = [];
  private $dbHandle;
  private $database;
  private $table;

  public function __construct($iniArray) 
  {
    $this->processOptions($iniArray);
    return true;
  }

  private function processOptions($iniArray)
  {
    $this->dbCredentialsIni = $iniArray['db_ini'];
    $this->database = $iniArray['database'];
    $this->table = $iniArray['table'];
    return true;
  }

  public function setEmailFileNames() 
  {
    $cmd = "ls -1d {$this->emailDirectory}/*";
    system($cmd,$retval);
    if ($retval == 2) {
      return false;
    }
    $emails =  explode("\n",`$cmd`);
    array_pop($emails);
    $this->emailFileNames = $emails;
    return true;
  }

  public function readEmails() {
  $emailContents = [];
  foreach ($this->emailFileNames as $fileName) {
    $emailContents[$fileName] =  file_get_contents($fileName);
    $this->copyEmailFiles($fileName);
    $this->deleteEmailFiles($fileName);
  }
  $this->emailContents = $emailContents;
  return true;
  }

  private function deleteEmailFiles($file)
  {
    system("rm -vf $file");
    return true;
  }

  private function copyEmailFiles($file)
  {
    $justFileName = str_replace("{$this->emailDirectory}/",'',$file);
    system("mkdir -p /tmp/save");
    system("cp -vf $file /tmp/save/{$justFileName}");
    return true;
  }

  public function parseEmails() {
    foreach ($this->emailContents as $fileName => $emailData)
    {
      $this->parseEmailData($emailData);
    }
    return true;
  }

  private function parseEmailData($emailData) {
    $emailDataExploded = [];
    $emailDataExploded = explode("\n",$emailData);
    $emailParsedData = []; 
    foreach($emailDataExploded as $lineNumber => $lineData) 
    {
	if (preg_match("/^Subject:/",$lineData)) {
	  $nextLine = $lineNumber + 1;
	  $emailParsedData['subject'] =  $emailDataExploded[$lineNumber] . $emailDataExploded[$nextLine];
	}
	if (preg_match("/^Date:/",$lineData)) {
	  $nextLine = $lineNumber + 1;
	  $emailParsedData['date'] =  $emailDataExploded[$lineNumber] . $emailDataExploded[$nextLine];
	}
    }
    $this->emailParsedData[] = $emailParsedData;
    return true;
  }

  private function connectToDB()
  {
    // Open connection
    $dbCredentials = parse_ini_file($this->dbCredentialsIni);
    try 
    {
	    $pdo = new PDO('mysql:host=localhost;dbname=backups', $dbCredentials['user'], $dbCredentials['password']);

    }
    catch (PDOException $e) 
    {
	echo 'Error: ' . $e->getMessage();
	exit();
    }
    $this->dbHandle = $pdo;
    return true;
  }

  public function updateDatabase()
  {
    $this->connectToDB();
    foreach ($this->emailParsedData as $arrayNumber => $parsedArray) 
    {
      $parsedSubject = $this->parseSubject($parsedArray['subject']);
      $parsedDate = $this->parseDate($parsedArray['date']);
      $stmt = $this->generateReplaceStatements($parsedSubject,$parsedDate);
      try 
      {
	$this->dbHandle->exec($stmt);
      }
      catch (PDOException $e) 
      {
	  echo 'Error: ' . $e->getMessage();
      }
    }
    return true;
  }

  private function generateReplaceStatements($parsedSubject,$parsedDate) {

    $date = $parsedDate->format('Y-m-d H:i:s');
    $replaceStatement = "REPLACE INTO {$this->database}.{$this->table} ";
    $replaceStatement .= "(client,server,company,code,backupdate,date_modified) VALUES ";
    $replaceStatement .= "(" . $this->dbHandle->quote($parsedSubject['client']) . "," ;
    $replaceStatement .= $this->dbHandle->quote($parsedSubject['server']) . "," ;
    $replaceStatement .= $this->dbHandle->quote($parsedSubject['company']) . "," ;
    $replaceStatement .= $this->dbHandle->quote($parsedSubject['code']) . "," ;
    $replaceStatement .= "'$date',NOW())";
    return $replaceStatement;
  }

  private function parseSubject($subject)
  {
    $subject = preg_replace("/^Subject:/",'',$subject);
    LIST($client,$server,$company,$skip1,$skip2,$code) = explode("|",$subject);
    $code = preg_replace("/[^0-9]/",'',$code);
    $parsedSubject = ['client' => trim($client), 'server' => trim($server), 'company' => trim($company), 'code' => trim($code)];
    return $parsedSubject;
  }

  private function parseDate($date) 
  {
    preg_match('/\s\d+\s[A-Z,a-z]{3,3}\s[0-9]{4,4}\s[0-9]{2,2}:[0-9]{2,2}:[0-9]{2,2}/', $date, $matches, PREG_OFFSET_CAPTURE);
    LIST($day,$month,$year,$time) = explode(' ',trim($matches[0][0]));
    $month = $this->convertMonthToNumber($month);
    $dateProcessed = "$year-$month-$day $time";
    $dt = new DateTime($dateProcessed);
    $dt->setTimezone(new DateTimeZone('US/Eastern'));
    return $dt;
  }

  private function convertMonthToNumber($month)
  {
    $months = [
    'Dec' => 12,
    'Nov' => 11,
    'Oct' => 10,
    'Sep' => 9,
    'Aug' => 8,
    'Jul' => 7,
    'Jun' => 6,
    'May' => 5,
    'Apr' => 4,
    'Mar' => 3,
    'Feb' => 2,
    'Jan' => 1
    ];

    return $months[$month];
  }

}


checkThresholdsClass.php
<?php

class checkThresholds
{

  private $dbCredentialsIni;
  private $thresholds = [];
  private $thresholdsCleanedUp = [];
  private $dbHandle;
  private $database;
  private $table;
  private $dbData = [];
  private $mergedThresholds = [];
  private $reportData = [];
  private $cliReport;
  private $webPageFile; 
  private $stateCount = [ 'OK' => 0 , 'UNKNOWN' => 0, 'CRITICAL' => 0, 'TOTAL' => 0];
  private $newEntriesForIni = [];
  private $defaultThreshold; 

  public function __construct($iniArray) 
  {
    $this->processOptions($iniArray);
    return true;
  }

  private function processOptions($iniArray)
  {
    $this->dbCredentialsIni = $iniArray['db_ini'];
    $this->thresholds = $iniArray['threshold'];
    $this->database = $iniArray['database'];
    $this->table = $iniArray['table'];
    $this->webPageFile = $iniArray['webPageFile'];
    $this->defaultThreshold = $iniArray['defaultThreshold'];
    return true;
  }

  public function determineStatus()
  {
    $this->dbData = $this->readDB();
    $this->thresholdsCleanedUp = $this->cleanupThresholdArray();
    $this->mergedThresholds = $this->mergeThresholdsWithDbData();
    $this->reportData = $this->generateStatusReportData();
    return true;
  }

  public function generateCLIreport()
  {
    $cliReport = '';
    foreach($this->reportData as $uniqKey => $rData)
    {
      if ($rData['emailReceived'] === TRUE) {
	if ($rData['overdue'] === TRUE) {
	  $cliReport .= "UNKOWN: Backup Email Overdue: Company: {$rData['data']['company']} Client: {$rData['data']['client']} Last Email: {$rData['data']['dbResults']['backupdate']}";
	} else {
	  if ($rData['data']['dbResults']['code'] == 1121) {
	    $cliReport .= "CRITCAL: Backup Failed: Company: {$rData['data']['company']} Client: {$rData['data']['client']} Code:{$rData['data']['dbResults']['code']} Last Email: {$rData['data']['dbResults']['backupdate']}";
	  } else {
	    $cliReport .= "OK: Backup Successful: Company: {$rData['data']['company']} Client: {$rData['data']['client']} Code:{$rData['data']['dbResults']['code']} Last Email: {$rData['data']['dbResults']['backupdate']}";
	  }
	}
      } else {
	$cliReport .= "UNKOWN: Email Never Received: Company: {$rData['data']['company']} Client: {$rData['data']['client']} ";
      }

      $cliReport .= "\n";
    }
    echo "$cliReport";
  }

  public function generateWebPage()
  {
    $webPageData = $this->generateWebPageData();
    $this->generateWebPageTable($webPageData); 
    return true;
  }

  private function generateWebPageTable($webPageData)
  {
    $table = "<table border=1>\n";
    $table .= "<tr><th> id </th> <th> Status </th> <th> Cause </th> <th> Company </th> <th> Backup Client </th> <th> Backup Code </th> <th> Last Email </th> <th> Threshold-Hours </th><th> Connect </th> </tr>\n";
    $rowArray = []; 
    foreach($webPageData as $state => $cause)
    {
      $bgcolor = $this->getRowColor($state);
      foreach ($cause as $causeName => $data)
      {
	foreach ($data as $dataProperties => $dp) 
	{
	  $rows = '';
	  $rows  .= "<tr bgcolor=$bgcolor>";
	  $rows  .= "<td>idreplace</td> ";
	  $rows  .= "<td align=center>$state</td> ";
	  $rows  .= "<td align=center>$causeName</td> ";
	  $rows  .= "<td align=center>{$dp['data']['company']}</td> ";
	  if (isset($dp['data']['notInIni']) && $dp['data']['notInIni'] == TRUE) {
	    $rows  .= "<td align=center>{$dp['data']['client']} (not in ini)</td> ";
	  } else {
	    $rows  .= "<td align=center>{$dp['data']['client']}</td> ";
	  }
	  if (isset($dp['data']['dbResults']['code'])) {
		  $rows  .= "<td align=center>{$dp['data']['dbResults']['code']}</td> ";
	  } else {
		  $rows  .= '<td>unknown </td>';
	  }
	  if (isset($dp['data']['dbResults']['backupdate'])) {
		  $rows  .= "<td>{$dp['data']['dbResults']['backupdate']}</td> ";
	  } else {
		  $rows  .= '<td> unknown </td>';
	  }
	  $rows  .= "<td align=center>{$dp['data']['hour']}</td> ";
	  if (!empty($dp['data']['url']))
	  {
		    $rows  .= "<td align=center><a href='{$dp['data']['url']}' target='_blank'> Connect </a></td> ";
	  } else {
		    $rows  .= "<td align=center></td> ";
	  }
	  $rows .= "</tr>\n";
	  $sortOrder = $this->sortOrder($state);
	  $this->incrementStateCount($state);
	  $rowsArray[$sortOrder][] = $rows;
	}
      }
    }
    sort($rowsArray);
    $id = 1;
    foreach ($rowsArray as $order => $orderNumber) {
      foreach ($orderNumber as $tableRow) {
	$table .= str_replace('idreplace',$id,$tableRow);
	$id++;
      }
    }
    $table .= "</table>";
    $webPage = '';
    $webPage .= "<html> <head> <title> NLS Backups </title> </head><body>\n";
    $currentTimestamp = new DateTime();
    $currentTimestamp->setTimezone(new DateTimeZone('US/Eastern'));
    $webPage .= "This page was last updated at " . $currentTimestamp->format("Y-m-d H:i:s") . " Eastern Time <br>";
    $dashboard = $this->generateDashBoardStateCounts();
    $webPage .= $dashboard;
    $webPage .= $table;
    $webPage .= "</body></html>\n";
    file_put_contents($this->webPageFile, $webPage);
    return true; 
  }

  private function getRowColor($state) {
    $colors = [
      'UNKNOWN' => 'ORANGE',
      'CRITICAL' => 'E64A34',
      'OK' => 'WHITE',
      'TOTAL' => 'WHITE'
      ];
      return $colors[$state];
  }

  private function sortOrder($state) {
    $order = [
      'UNKNOWN' => 1,
      'CRITICAL' => 2,
      'OK' => 3
      ];
      return $order[$state];
  }

  private function incrementStateCount($state) {
    $this->stateCount['TOTAL'] = $this->stateCount['TOTAL'] + 1;
    $this->stateCount[$state] = $this->stateCount[$state] + 1;
    return true;
  }

  private function generateDashBoardStateCounts()
  {
    $table = "<table border=1>";
    $row = "<tr>";
    foreach ($this->stateCount as $state => $count) {
      $bgcolor = $this->getRowColor($state);
      $row .= "<td bgcolor=$bgcolor> $state </td>";
    }
      $row .= "</tr>";
    foreach ($this->stateCount as $state => $count) {
      $bgcolor = $this->getRowColor($state);
      $row .= "<td bgcolor=$bgcolor align=center> $count </td>";
    }
      $row .= "</tr>";
      $table .= $row;
      $table .= "</table>";
      return $table; 
  }

  private function generateWebPageData()
  {
    $webPageData = [];
    foreach($this->reportData as $uniqKey => $rData)
    {
      if ($rData['emailReceived'] === TRUE) {
	if ($rData['overdue'] === TRUE) {
	  $webPageData['UNKNOWN']['overdue'][] = $rData;
	} else {
	  if ($rData['data']['dbResults']['code'] == 1121) {
	    $webPageData['CRITICAL']['failed'][] = $rData;
	  } else {
	    $webPageData['OK']['successful'][] = $rData;
	  }
	}
      } else {
	$webPageData['UNKNOWN']['missing'][] = $rData;
      }
    }
    return $webPageData;
  }



  private function generateStatusReportData()
  {
    $report = [];
    foreach ($this->mergedThresholds as $backupJob => $thresh) 
    {
      $key = base64_encode($thresh['client'] . '-' . $thresh['server'] . '-' . $thresh['company']);
      $report[$key]['data'] = $thresh;
      $thresholdTimestamp = $this->calculateThresholdDate($thresh['hour']);
      if (!isset($thresh['dbResults'])){
	$report[$key]['emailReceived'] = FALSE;
      } else {
	$report[$key]['emailReceived'] = TRUE;
      }

      if (isset($thresh['dbResults']))
      {
	if ($thresh['dbResults']['backupdate'] < $thresholdTimestamp->format("Y-m-d H:i:s")) {
	  $report[$key]['overdue'] = TRUE;
	} else {
	  $report[$key]['overdue'] = FALSE;
	}
      }
    }
    
    return $report;
  }

  private function calculateThresholdDate($hours) {
    $currentTimestamp = new DateTime();
    $currentTimestamp->setTimezone(new DateTimeZone('US/Eastern'));
    $currentTimestamp->sub(new DateInterval("PT{$hours}H"));
    return $currentTimestamp;
  }

  private function mergeThresholdsWithDbData()
  {
    $dataFromDB = $this->dbData;
    foreach ($this->thresholdsCleanedUp as $key => $thresholds)
    {
      foreach ($dataFromDB as $row => $dbData) 
      {
	if ($thresholds['client'] == $dbData['client'] && $thresholds['server'] == $dbData['server'] && $thresholds['company'] == $dbData['company']) {
	  $this->thresholdsCleanedUp[$key]['dbResults'] = $dbData;
	  unset($dataFromDB[$row]);
	}
      }
    }
    $this->newEntriesForIni = $dataFromDB;
    $mergedThresholds  =  $this->processNewBackupJobs($this->thresholdsCleanedUp,$this->newEntriesForIni,$this->defaultThreshold);
    return $mergedThresholds;
  }

  private function processNewBackupJobs($thresholdCleanUp,$newEntries,$threshold)
  {
  foreach ($newEntries as $row => $dbData)
    {
      $newEntry = [];
      $newEntry['client'] = $dbData['client'];
      $newEntry['server'] = $dbData['server'];
      $newEntry['company'] = $dbData['company'];
      $newEntry['hour'] = $threshold;
      $newEntry['notInIni'] = TRUE;
      $newEntry['dbResults'] = $dbData;
      $thresholdCleanUp[] = $newEntry;
    }
    return $thresholdCleanUp;
  }

  private function readDB()
  {
    $dbData = [];
    $this->connectToDB();
    $query = "SELECT * from {$this->database}.{$this->table} where code <> ''";
    $results = $this->dbHandle->query($query,PDO::FETCH_ASSOC);
    foreach ($results as $row => $values) {
      $dbData[] = $values;
    }
    return $dbData;
  }

  private function cleanupThresholdArray()
  {
    $thresholdsCleaned = [];
    foreach ($this->thresholds as $threshold) {
      LIST($client,$server,$company,$hour,$url) = explode("|",$threshold);
	$thresholdsCleaned[] = [ 
	  'client' => trim($client),
	  'server' => trim($server),
	  'company' => trim($company),
	  'hour' => trim($hour),
	  'url' => trim($url)
	  ];
    }
    return $thresholdsCleaned;
  }

  private function connectToDB()
  {
    // Open connection
    $dbCredentials = parse_ini_file($this->dbCredentialsIni);
    try 
    {
	    $pdo = new PDO('mysql:host=localhost;dbname=backups', $dbCredentials['user'], $dbCredentials['password']);

    }
    catch (PDOException $e) 
    {
	echo 'Error: ' . $e->getMessage();
	exit();
    }
    $this->dbHandle = $pdo;
    return true;
  }


}


parse_shadow_backup_emails.py
import re
from datetime import datetime
import json
import glob
import os
import pendulum


class EmailFiles:
    @staticmethod
    def get_list_of_files(pattern):
        return glob.glob(pattern)


class ParseShadowBackupEmails:
    active_email_dictionary = {}
    web_page_file = 'sp.html'
    master_email_dictionary = {}
    change_count = 0
    backup_code_1120 = []
    backup_code_1121 = []
    backup_code_unknown = []
    default_threshold = 48
    row_id = 0
    date_format = '%m-%d-%Y %H:%M:%S'
    max_parse_time = 'none'
    min_parse_time = 'none'
    company_count = 0
    table_width = "100%"

    def __init__(self, directory):
        self.dictionary_file = os.getenv("HOME") + '/email_dictionary.json'
        list_of_email_files = self.check_for_shadow_email(EmailFiles.get_list_of_files(directory))
        subjects = self.get_subjects(list_of_email_files)
        split_subjects = self.split_subject(subjects)
        self.build_unique_active_dictionary(split_subjects)
        self.load_master_dictionary()
        self.compare_master_and_active_dictionaries()

    @staticmethod
    def check_for_shadow_email(list_of_files):
        list_of_shadow_emails = []
        for email in list_of_files:
            with open(email) as f:
                if 'ShadowProtectSvc' in f.read():
                    list_of_shadow_emails.append(email)
        return list_of_shadow_emails

    def get_subjects(self, list_of_email_files):
        subjects = []
        for email in list_of_email_files:
            with open(email) as f:
                file_data = f.readlines()
                subjects.append(self.get_match_and_next_line("^Subject:", file_data))
                os.remove(email)
        return subjects

    def rename_web_page(self):
        os.rename(self.web_page_file, "/var/www/html/reports/" + self.web_page_file)

    @staticmethod
    def get_match_and_next_line(pattern, file_data):
        from_date = ''
        for line_number, line in enumerate(file_data):
            if re.search("^From", line) and line_number == 1:
                from_date = "| " + line
            if re.search(pattern, line):
                subject = line.rstrip() + " " + file_data[line_number + 1].rstrip() + " " + from_date.rstrip()
                return subject

    def split_subject(self, subjects):
        split_subjects = []
        for subject in subjects:
            split_subject = subject.split('|')
            split_subjects.append(self.create_dictionary(split_subject))
        return split_subjects

    def create_dictionary(self, split_subject):
        if len(split_subject) == 7:
            subject_dictionary = {
                'server': str(split_subject[0].strip().replace('Subject: ', '')),
                'client': str(split_subject[1].strip()),
                'company': str(split_subject[2].strip()),
                'backup_code': str(self.get_backup_code(split_subject[5].strip())),
                'email_time': self.convert_utc_to_local(str(self.get_email_time(split_subject[-1].strip()))),
                'threshold': self.default_threshold
            }
            return subject_dictionary

    def convert_utc_to_local(self, email_time):
        dt = pendulum.from_format(email_time, 'MMM DD HH:mm:ss YYYY', tz='UTC')
        tz = pendulum.timezone('US/Eastern')
        in_est = tz.convert(dt)
        return in_est.strftime(self.date_format)

    def build_unique_active_dictionary(self, split_subjects):
        split_subjects = list(filter(None.__ne__, split_subjects))
        for subject in split_subjects:
            key = subject['company'] + subject['server'] + subject['client']
            if key not in self.active_email_dictionary:
                self.active_email_dictionary[key] = subject
            elif key in self.active_email_dictionary:
                datetime_object_in_dict = datetime.strptime(self.active_email_dictionary[key]['email_time'],
                                                            self.date_format)
                datetime_object = datetime.strptime(subject['email_time'], self.date_format)
                if datetime_object > datetime_object_in_dict:
                    self.active_email_dictionary[key] = subject
        self.initialize_dictionary()

    def compare_master_and_active_dictionaries(self):
        for key, email_data in self.active_email_dictionary.items():
            if key not in self.master_email_dictionary:
                self.master_email_dictionary[key] = email_data
                self.change_count += 1
            elif key in self.master_email_dictionary:
                datetime_object_in_master_dict = datetime.strptime(self.master_email_dictionary[key]['email_time'],
                                                                   self.date_format)
                datetime_object = datetime.strptime(email_data['email_time'], self.date_format)
                if datetime_object > datetime_object_in_master_dict:
                    if self.master_email_dictionary[key]['threshold'] != email_data['threshold']:
                        email_data['threshold'] = self.master_email_dictionary[key]['threshold']
                    self.master_email_dictionary[key] = email_data
                    self.change_count += 1
        if self.change_count > 0:
            print("Something Change Updating Dictionary on disk")
            self.save_json(self.master_email_dictionary, self.dictionary_file)

    def generate_html_table(self):
        self.sort_master_dictionary_for_web_page()
        table = ''
        table += self.build_html_table_header()
        table += self.build_html_table_data('unknown', self.backup_code_unknown)
        table += self.build_html_table_data('critical', self.backup_code_1121)
        table += self.build_html_table_data('ok', self.backup_code_1120)
        table += self.build_html_table_footer()
        with open(self.web_page_file, "w") as text_file:
            text_file.write(table)

    def build_dashboard(self):
        current_time = pendulum.now('US/Eastern')
        table = ''
        table += "<table border=1 width=" + self.table_width + ">\n"
        table += "<tr>\n"
        table += "<th> OK </th>\n"
        table += "<th> UNKNOWN </th>\n"
        table += "<th> CRITICAL </th>\n"
        table += "<th> Clients </th>\n"
        table += "<th> Companies </th>\n"
        table += "<th> First Email </th>\n"
        table += "<th> Last Email </th>\n"
        table += "<th> Page Generated </th>\n"
        table += "</tr>\n"
        table += "<tr>\n"
        table += "<td align=center> " + str(len(self.backup_code_1120)) + "</td>\n"
        table += "<td bgcolor=orange align=center> " + str(len(self.backup_code_unknown)) + "</td>\n"
        table += "<td bgcolor=E64A34 align=center> " + str(len(self.backup_code_1121)) + "</td>\n"
        table += "<td align=center> " + str(
            len(self.backup_code_1121) + len(self.backup_code_1120) + len(self.backup_code_unknown)) + "</td>\n"
        table += "<td align=center> " + str(self.company_count) + "</td>\n"
        table += "<td align=center> " + self.min_parse_time.strftime(self.date_format) + "</td>\n"
        table += "<td align=center> " + self.max_parse_time.strftime(self.date_format) + "</td>\n"
        table += "<td align=center> " + current_time.strftime(self.date_format) + "</td>\n"
        table += "</tr></table>\n"
        return table

    def build_html_table_data(self, status, list_of_code):
        table_data = ''
        for data in list_of_code:
            self.row_id += 1
            table_data += self.build_table_row_data(self.row_id, status, data)
        return table_data

    @staticmethod
    def get_row_color(status):
        colors = {'unknown': 'orange', 'ok': 'white', 'critical': 'E64A34'}
        return colors[status]

    def build_table_row_data(self, row_id, status, row):
        table_row = ''
        table_row += "<tr bgcolor=" + self.get_row_color(status) + ">"
        table_row += "<td align=center>" + str(row_id) + "</td>"
        table_row += "<td align=center>" + status.upper() + "</td>"
        table_row += "<td align=center>" + row['company'] + "</td>"
        table_row += "<td align=center>" + row['client'] + "</td>"
        table_row += "<td align=center>" + row['email_time'] + "</td>"
        table_row += "<td align=center>" + row['backup_code'] + "</td>"
        table_row += "<td align=center>" + str(row['threshold']) + " Hours</td>"
        table_row += "</tr>\n"
        return table_row

    def build_html_table_header(self):
        header = ['id', 'status', 'company', 'client', 'last email', 'backup_code', 'threshold']
        table_head = ''
        table_head += "<html>\n"
        table_head += "<head>\n"
        table_head += "<title> Shadow Protect Backup Status </title>\n"
        table_head += "</head>\n"
        table_head += self.build_dashboard()
        table_head += "<table border=1 width=" + self.table_width + ">\n"
        table_head += "<tr>"
        for column in header:
            table_head += '<th>' + column.upper() + '</th>'
        table_head += "</tr>\n"
        return table_head

    @staticmethod
    def build_html_table_footer():
        footer = ''
        footer += "</table>"
        footer += "</html>"
        return footer

    def sort_master_dictionary_for_web_page(self):
        company_list = []
        for key, data in sorted(self.master_email_dictionary.items()):
            company_list.append(data['company'])
            self.get_parse_time(data['email_time'])
            now = pendulum.now('US/Eastern')
            threshold_time = now.subtract(hours=data['threshold'])
            alert_time_formatted = threshold_time.strftime(self.date_format)
            alert_time = datetime.strptime(alert_time_formatted, self.date_format)
            email_time = datetime.strptime(data['email_time'], self.date_format)
            if email_time < alert_time:
                self.backup_code_unknown.append(data)
            elif data['backup_code'] == "1120":
                self.backup_code_1120.append(data)
            elif data['backup_code'] == "1121":
                self.backup_code_1121.append(data)
        self.company_count = len(set(company_list))

    def get_parse_time(self, parse_time):
        if self.max_parse_time == 'none':
            self.max_parse_time = datetime.strptime(parse_time, self.date_format)
        elif datetime.strptime(parse_time, self.date_format) > self.max_parse_time:
            self.max_parse_time = datetime.strptime(parse_time, self.date_format)
        if self.min_parse_time == 'none':
            self.min_parse_time = datetime.strptime(parse_time, self.date_format)
        elif datetime.strptime(parse_time, self.date_format) < self.min_parse_time:
            self.min_parse_time = datetime.strptime(parse_time, self.date_format)

    def load_master_dictionary(self):
        if os.path.isfile(self.dictionary_file):
            self.master_email_dictionary = self.load_json(self.dictionary_file)

    def initialize_dictionary(self):
        if not os.path.isfile(self.dictionary_file):
            print("Initialized Dictionary on Disk")
            try:
                self.save_json(self.active_email_dictionary, self.dictionary_file)
            except (FileNotFoundError, FileExistsError, IOError) as err:
                print(err)

    @staticmethod
    def save_json(data, file_name):
        try:
            with open(file_name, 'w') as fp:
                json.dump(data, fp)
        except (FileNotFoundError, FileExistsError, IOError) as err:
            print(err)

    @staticmethod
    def load_json(file_name):
        try:
            with open(file_name, 'r') as fp:
                return json.load(fp)
        except (FileExistsError, IOError) as err:
            print(err)

    @staticmethod
    def get_backup_code(raw_code):
        backup_code = raw_code.split()
        return backup_code[1].strip("'")

    @staticmethod
    def get_email_time(raw_email_time):
        email_time_list = raw_email_time.split()
        email_time = email_time_list[-4] + ' ' + email_time_list[-3] + ' ' + email_time_list[-2] + ' ' + \
                     email_time_list[-1]
        return email_time
